### PR TITLE
Expose 'other' property and event accessors

### LINKED ISSUE #4100: Missing accessors on PropertyDefinition
The implementation of System.Reflection.Metadata [PropertyDefinition](https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/PropertyDefinition.cs#L91) and [EventDefinition](https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/EventDefinition.cs#L84) swallows some accessors and doesn't make them available publically. This has been marked as _TODO_ in the source code. These _'other'_ accessors are not often encountered, but handling them is required for certain special case tooling.

## usage of _'other'_ property accessors
* The .NET Framework exposes 'other' accessors through [PropertyInfo.GetAccessors](https://referencesource.microsoft.com/#mscorlib/system/reflection/propertyinfo.cs,496)
* .NET assemblies generated by the 'tlbimp' tool can generate this metadata when a COM property has both a PUT and a PUTREF accessor.
* They are valid metadata, even though rarely used, but if you build tools for processing or inspecting .NET assemblies (like tlbexp or ildasm do) you want to handle them.

The last point is a real-world use case I have. The tlbexp functionality built into the .NET Framework requires loading an assembly into the AppDomain to generate a TLB. Loading the assembly can fail for various reasons, so generating a TLB by only inspecting metadata is desireable. The omission of 'other' accessors is a problem here.

## Proposed API
``` csharp
namespace System.Reflection.Metadata
{
public struct PropertyAccessors 
{
    // Existing
    public MethodDefinitionHandle Getter { get; }
    public MethodDefinitionHandle Setter { get; } 

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}

public struct EventAccessors
{  
    // Existing
    public MethodDefinitionHandle Adder { get; }
    public MethodDefinitionHandle Remover { get; }
    public MethodDefinitionHandle Raiser { get; }

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}
}
```

## Details

In the large majority of cases there will be no 'other' accessors in the metadata. It is trivially possible to implement this case without any additional allocations compared to the current implementation.

For the purpose of the discussion I have provided an initial [implementation](https://github.com/weltkante/corefx/pull/3) of the suggested change.

## Updates

- Expose other accessors directly on the `PropertyAccessors` and `EventAccessors` structs instead of returning them as an out-param

### PR DESCRIPTION
**Pull request by weltkante**
Implementation of API to expose 'other' property and event accessors.

resolves #4100

### CODE REVIEW

**Comment by nguerrera** [comment]
Looks good to me. @tmat?

File: src/System.Reflection.Metadata/tests/Metadata/MetadataReaderTests.cs [hunk 1]
```diff
@@ -2432,6 +2432,39 @@ public void Bug17109()
...
+            var eventAccessors = eventDef.GetAccessors();
+            var otherAccessorNames = (from methodHandle in eventAccessors.Others
+                                      select reader.GetString(reader.GetMethodDefinition(methodHandle).Name)).ToArray();
+
+            Assert.Equal("add_Notification", reader.GetString(reader.GetMethodDefinition(eventAccessors.Adder).Name));
+            Assert.Equal("remove_Notification", reader.GetString(reader.GetMethodDefinition(eventAccessors.Remover).Name));
+            Assert.True(eventAccessors.Raiser.IsNil);
+
+            // ilasm doesn't retain the order in which other accessors were specified in IL
+            AssertEx.SetEqual(otherAccessorNames, "suspend_Notification", "resume_Notification", "other_Notification");
```
Comment (tmat) [inline src/System.Reflection.Metadata/tests/Metadata/MetadataReaderTests.cs]: >AssertEx [](start = 12, length = 8)

Nit: even if ilasm doesn't retain order we have the dll checked in, so we can test the exact order. If the dll is regenerated then the test might need to be updated but that's ok.

**Comment by tmat** [comment]
LGTM modulo testing the ordering.

**Comment by weltkante** [comment]
Failed test doesn't seem related to my code, says `Known race in .NET's EventListener in 4.6.1 and below. Retry the job or upgrade to 4.6.2` - not sure how to retry it though

**Comment by tmat** [comment]
test Innerloop netfx Windows_NT Debug x86 Build and Test  please

**Comment by tmat** [comment]
@weltkante Thanks a lot for your contribution.

**Comment by nguerrera** [comment]
Thanks, @weltkante !

**Pull request merged**